code/game/gamemodes/miniantags/pulsedemon/pulsedemon.dm

```diff
@@ -0,0 +1,841 @@
+	var/list/apc_images = list()
+	/// List of all previously hijacked APCs.
+	var/list/hijacked_apcs = list()
+	/// Reference to the APC currently being hijacked.
+	var/obj/machinery/power/apc/apc_being_hijacked
+	/// This demon's progressbar helper, used for the hijacking timer on APCs and all charger types (cell, gun, cyborg).
+	var/datum/progressbar_helper/pb_helper
+
+/mob/living/simple_animal/pulse_demon/Initialize(mapload)
+	. = ..()
+	if(!mapload)
+		name += " ([rand(100, 999)])"
+		real_name = name
+
+	remove_from_all_data_huds()
+	ADD_TRAIT(src, TRAIT_AI_UNTRACKABLE, PULSEDEMON_TRAIT)
+	flags_2 |= RAD_NO_CONTAMINATE_2
+
+	// don't step on me
+	RegisterSignal(src, COMSIG_CROSSED_MOVABLE, PROC_REF(try_cross_shock))
+	RegisterSignal(src, COMSIG_MOVABLE_CROSSED, PROC_REF(try_cross_shock))
+
+	// drop demon onto ground if its loc is a non-turf and gets deleted
+	RegisterSignal(src, COMSIG_PARENT_PREQDELETED, PROC_REF(deleted_handler))
+
+	RegisterSignal(SSdcs, COMSIG_GLOB_CABLE_UPDATED, PROC_REF(cable_updated_handler))
+
+	pb_helper = new()
+
+	current_power = locate(/obj/machinery/power) in loc
+	// in the case that both current_power and current_cable are null, the pulsedemon will die the next tick
+	if(!current_power)
+		current_cable = locate(/obj/structure/cable) in loc
+	else
+		forceMove(current_power)
+	update_glow()
+	playsound(get_turf(src), 'sound/effects/eleczap.ogg', 30, TRUE)
+	give_spells()
+
+/mob/living/simple_animal/pulse_demon/proc/deleted_handler(src, force)
+	SIGNAL_HANDLER
+	// assume normal deletion if we're on a turf, otherwise deletion could be inherited from loc
+	if(force || isnull(loc) || isturf(loc))
+		return FALSE
+	// if we did actually die, simple_animal/death will set del_on_death to FALSE before calling qdel
+	if(!del_on_death)
+		return FALSE
+	exit_to_turf()
+	return TRUE
+
+/mob/living/simple_animal/pulse_demon/proc/cable_updated_handler(SSdcs, turf/T)
+	SIGNAL_HANDLER
+	if(cable_images[T])
+		var/list/turf_images = cable_images[T]
+		for(var/image/current_image in turf_images)
+			client?.images -= current_image
+		turf_images.Cut()
+	else
+		cable_images[T] = list()
+
+	for(var/obj/structure/cable/C in T)
+		var/image/cable_image = image(C, C, layer = ABOVE_LIGHTING_LAYER, dir = C.dir)
+		cable_image.plane = ABOVE_LIGHTING_PLANE
+		cable_images[T] += cable_image
+		client?.images += cable_image
+
+/mob/living/simple_animal/pulse_demon/proc/apc_deleted_handler(obj/machinery/power/apc/A, force)
+	SIGNAL_HANDLER
+	hijacked_apcs -= A
+
+/mob/living/simple_animal/pulse_demon/Destroy()
+	pb_helper.cancel() // just making sure nothing started between death() and here
+	QDEL_NULL(pb_helper)
+
+	QDEL_LIST_CONTENTS(cable_images)
+	QDEL_LIST_CONTENTS(apc_images)
+
+	controlling_area = null
+	current_bot = null
+	current_cable = null
+	current_power = null
+	current_robot = null
+	current_weapon = null
+	apc_being_hijacked = null
+	hijacked_apcs = null
+	hijacked_robots = null
```

Comment:
all of these things should probably be UIDs except maybe the `current` vars and the controlling area var